content/chapters/compute/lab/README.md

```diff
@@ -282,6 +282,201 @@ Therefore, there is also a gray area where there is no clear right answer to thi
 One such gray area are browsers.
 Some open a new tab in a separate process, while others do it in a separate thread instead.
 
+## Usage of processes and threads in apache2
+
+We'll take a look at how a real-world application - the apache2 http server - makes use of processes and threads.
+Since the server must be able to handle multiple clients at the same time, it must therefore use some form of concurrency: when a new client arrives, the server offloads the work of interacting with that client to another process or thread.
+
+The choice of whether to use multiple processes or threads is not baked in into the code. Instead, apache2 provides a couple of modules called MPMs (Multi-Processing Modules). Each module implements a different concurrency model and the users can pick whatever module better fits their needs by editing the server configuration files.
+
+The most common MPMs are
+- `prefork`: there are multiple worker processes, each process is single-threaded and handles one client request at a time
+- `worker`: there are multiple worker processes, each process is multi-threaded, and each thread handles one client request at a time
+- `event`: same as `worker` but designed to better handle some particular use cases
+
+In principle, `prefork` provides more stability and backwards compatibility, but it has a bigger overhead. On the other hand, `worker` and `event` are more scalable, and thus able to handle more simultaneous connections, due to the usage of threads. On modern systems `event` is almost always the default.
+
+### apache2 live action
+
+Let's run an actual instance of apache2 and see how everything works. Go to `support/apache2` and run `make run`. This will start a container with apache2 running inside.
+
+Check that the server runs as expected
+
+```Bash
+student@os:~$ curl localhost:8080
+<html><body><h1>It works!</h1></body></html>
+```
+
+Now go inside the container and take a look at running processes:
+
+```Bash
+student@os:~/.../lab/support/apache2$ docker exec -it apache2-test bash
+root@56b9a761d598:/usr/local/apache2# ps -ef
```

Comment:
```suggestion

root@56b9a761d598:/usr/local/apache2# ps -ef
```
Leave one empty line between subsequent commands so they're easier to tell apart.